File: Array&String/HashTable/rank_transform_of_an_array.py
class Solution:
    def arrayRankTransform(self, arr: list[int]) -> list[int]:
        temp, rank, cnt, ans = sorted(arr), {}, 1, []
        for x in temp:
            if rank.get(x): continue
            rank[x] = cnt
            cnt += 1

        for x in arr: ans.append(rank[x])
        return ans

    # shorter
    def arrayRankTransform(self, arr: list[int]) -> list[int]:
        temp = sorted(set(arr))
        rank = {value : i+1 for i, value in enumerate(temp)}
        return [rank[x] for x in arr]

File: Array-String/HashTable/CPP/rank_transform_of_an_array.cpp
#include <algorithm>
#include <map>
#include <set>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        vector<int> temp = arr;
        sort(temp.begin(), temp.end());
        int cnt = 1;
        map<int, int> rank;
        for (int x: temp) {
            if (rank.count(x) == 1) continue;
            rank[x] = cnt;
            cnt++;
        }

        vector<int> ans;
        for (int x: arr) ans.push_back(rank[x]);
        return ans;
    }

    // without mapping
    vector<int> arrayRankTransform(vector<int>& arr) {
        set<int> rank(arr.begin(), arr.end());
        vector<int> temp(rank.begin(), rank.end());

        vector<int> ans;
        for (int x: arr) {
            auto it = find(temp.begin(), temp.end(), x);
            ans.push_back(it - temp.begin() + 1);
        }
        return ans;
    }
};